File: src/types/x64/ast.rs
#[derive(Debug)]
pub enum Program {
    Function(Function),
}

#[derive(Debug)]
pub struct Function {
    pub name: Identifier,
    pub body: Vec<Instruction>,
}

#[derive(Debug, Clone)]
pub struct Identifier(pub String);

#[derive(Debug)]
pub enum Instruction {
    Mov {
        // dst is always Operand::Stack.
        dst: Operand,
        src: Operand,
    },
    // Operand is always Operand::Stack.
    Unary(UnaryOp, Operand),
    Binary {
        op: BinaryOp,
        // dst is always Operand::Stack.
        dst: Operand,
        src: Operand,
    },
    Idiv(Operand),
    Cdq,
    AllocateStack(usize),
    Ret,
}

#[derive(Debug)]
pub enum UnaryOp {
    Neg,
    Not,
}

#[derive(Debug)]
pub enum BinaryOp {
    Add,
    Sub,
    Mul,
}

#[derive(Debug, Clone)]
pub enum Operand {
    Imm(i32),
    Reg(Reg),
    Pseudo(Identifier),
    Stack(usize),
}

#[derive(Debug, Clone)]
pub enum Reg {
    Ax,
    Dx,
    R10,
    R11,
}

File: lib/riscv64/ast.ml
open Stdint
open Common

type unary_op =
  | Neg
  | Not
[@@deriving show]

type reg =
  | A0
  | T0
  | T1
[@@deriving show]

type operand =
  | Imm of (uint64[@printer fun fmt v -> Format.fprintf fmt "%s" (Uint64.to_string v)])
  | Reg of reg
  | Stack of int
[@@deriving show]

type instruction =
  | Mov of
      { src : operand
      ; dst : operand
      }
  | Unary of
      { uop : unary_op
      ; src : operand
      ; dst : operand
      }
  | Li of
      { src : operand (* Imm *)
      ; dst : operand (* Reg *)
      }
  | Ld of
      { src : operand (* Stack *)
      ; dst : operand (* Reg *)
      }
  | Sd of
      { src : operand (* Reg *)
      ; dst : operand (* Stack *)
      }
  | AllocStack of int
  | Ret
[@@deriving show]

type function_def =
  | Function of
      { name : identifier
      ; body : instruction list
      }
[@@deriving show]

type program = Program of function_def list [@@deriving show]
